File: assets/ruby/tests/cleancode/test_no_provably_unnecessary_fallback.rb
# frozen_string_literal: true

# Test: no provably unnecessary fallback patterns.
# Detects x || y where x is a always-truthy literal.

require "minitest/autorun"
require "parser/current"

PROJECT_ROOT = Dir.pwd.freeze
SRC_DIRS = %w[src app lib].freeze
IGNORE_DIRS = %w[vendor node_modules dist build coverage].freeze

def source_files
  SRC_DIRS.flat_map do |dir|
    path = File.join(PROJECT_ROOT, dir)
    next [] unless File.directory?(path)

    Dir.glob(File.join(path, "**/*.rb")).select do |f|
      rel = Pathname.new(f).relative_path_from(Pathname.new(PROJECT_ROOT)).to_s
      IGNORE_DIRS.none? { |ig| rel.include?(ig) }
    end
  end
end

def definitely_truthy?(node)
  case node.type
  when :str then !node.children.first.empty?
  when :int, :float then node.children.first != 0
  when :true then true
  when :array, :hash then !node.children.first.to_a.empty?
  when :sym then true
  else false
  end
end

SOURCE_FILES = source_files.freeze

SOURCE_FILES.each do |filepath|
  define_method("test_no_unnecessary_fallback_in_#{filepath.gsub(/[^a-zA-Z0-9]/, '_')}") do
    code = File.read(filepath)
    buffer = Parser::Source::Buffer.new(filepath)
    buffer.source = code
    parser = Parser::CurrentRuby.new
    ast = parser.parse(buffer)
    skip "Cannot parse #{filepath}" unless ast

    violations = []
    walker = ->(node) do
      return unless node.is_a?(AST::Node)

      if node.type == :or && definitely_truthy?(node.children[0])
        violations << "  L#{node.loc.first_line}: left side of `or` is always truthy"
      end
      node.children.each { |c| walker.call(c) if c.is_a?(AST::Node) }
    end
    walker.call(ast)

    assert_empty violations, "Unnecessary fallback(s) in #{filepath}:\n#{violations.join("\n")}"
  end
end

File: assets/python/tests/cleancode/test_no_provably_unnecessary_fallback.py
"""Test: no provably unnecessary fallback patterns.

Detects ``or`` expressions where the left side is a literal that is
always truthy, making the right-hand side (fallback) dead code.
"""

import ast
from pathlib import Path

import pytest

PROJECT_ROOT = Path.cwd()
SRC_DIRS = ["src", "app", "lib"]
IGNORE_DIRS = {"venv", ".venv", ".env", "node_modules", "dist", "build", "coverage", "__pycache__"}


def _find_source_files() -> list[Path]:
    files = []
    for d in SRC_DIRS:
        src = PROJECT_ROOT / d
        if not src.is_dir():
            continue
        for path in src.rglob("*.py"):
            if not any(part in IGNORE_DIRS for part in path.relative_to(PROJECT_ROOT).parts):
                files.append(path)
    return files


def _is_definitely_truthy(node: ast.expr) -> bool:
    match node:
        case ast.Constant(value=str() as s):
            return len(s) > 0
        case ast.Constant(value=int() as n):
            return n != 0
        case ast.Constant(value=float() as n):
            return n != 0.0
        case ast.Constant(value=bool() as b):
            return b
        case ast.List(elts=elts):
            return len(elts) > 0
        case ast.Tuple(elts=elts):
            return len(elts) > 0
        case ast.Set(elts=elts):
            return len(elts) > 0
        case ast.Dict(keys=keys):
            return len(keys) > 0
        case _:
            return False


_SOURCE_FILES = _find_source_files()


@pytest.mark.parametrize("filepath", _SOURCE_FILES, ids=lambda p: str(p.relative_to(PROJECT_ROOT)))
def test_no_provably_unnecessary_fallback(filepath: Path) -> None:
    code = filepath.read_text(encoding="utf-8")
    try:
        tree = ast.parse(code)
    except SyntaxError:
        pytest.skip(f"Cannot parse {filepath}")

    violations: list[str] = []

    for node in ast.walk(tree):
        if not isinstance(node, ast.BoolOp) or node.op.__class__ is not ast.Or:
            continue
        if len(node.values) < 2:
            continue
        left = node.values[0]
        if _is_definitely_truthy(left):
            violations.append(f"  L{left.lineno}: left side of `or` is always truthy")

    assert not violations, f"Unnecessary fallback(s) in {filepath.relative_to(PROJECT_ROOT)}:\n" + "\n".join(violations)
